Lib/swiginit.swg

```diff
@@ -50,13 +50,29 @@ extern "C" {
 #define SWIGRUNTIME_DEBUG
 #endif
 
+static void SWIG_InitializeTypes(void);
```

Comment:
This is a generic file used by all languages and should not have any lanaguage specific code. The Javascript code needs moving to Javascript specific files.

Reply:
Yes, and it's common subroutine and it's the **same** for all languages. If you consider the corresponding https://github.com/swig/swig/pull/1973/commits/26f95e964c6b14acdde979611fad891364d5b7c5 by itself and **disregard** the `static class SWIG_PreInit` thing for a moment, you'll see that the overall result would be 100% equivalent to the original. So splitting per se is perfectly language-neutral. Now, it was also assumed that calling `SWIG_InitializeTypes` separately from static constructor would also be language-neutral. This turned to be wrong assumption, so I'll have to figured it out. For the moment I make static constructor dependent on SWIGJAVASCRIPT macro, but yes, it needs better solution. First I need to understand if the reasons for why it fails for, say, Python are applicable even to Javascript. If it can't happen in Javascript, then I'll simply move the static constructor to Javascript module...

Just in case for reference. Why static constructor? Since static constructor is executed at library load, and we know that library loading is MT-safe, we can use the latter property to avoid race conditions in `SWIG_InitializeModule`. It's attractive, because the approach is platform-neutral, i.e. you don't have to think if it's POSIX or Windows.

Reply:
Cool. I'd prefer initialisation to be done during the SWIG_InitializationModule if possible as initialization problems during file load are a pig to debug. A singleton function which returns a pointer to a static local instance can be initialized in a thread-safe manner when it is first called for C++11 and later, although in practice most compilers implement it as thread-safe for C++98 too.

Reply:
Call it occupational hazard, as I do have tendency to minimize reliance on compiler and run-time specifics and prefer solutions that are universal:-) In some degree it's a side effect of dealing with VC compiler, which to this day appears to be reluctant to commit to standards. `__cplusplus` pre-define is an example of that. They do let you specify `/std:c++14,17,latest` on compiler command line, but do they reflect it in `__cplusplus` pre-define? No, it reads `199711L` irregardless. I apologize for only tangentially related rant:-)

I've omitted the static constructor in favour of C++11 mutex. This effectively implies that those who intend to use swig-generated module with node.js worker_threads have to ask compiler for at least C++11 compliance. It's appropriate to inform the user about it. I can add a note to Doc/Manual/Javascript.html, but  one can also add `#warning` in `#else` condition for `<mutex>`. Would it be appropriate?

Reply:
> one can also add `#warning` in `#else` condition for `<mutex>`

Another possibility is to have an `assert` in `#else` condition for quard_locker that would try to detect if worker_threads are used at run-time and crash in a more controllable manner. This can be done by saving `isolate` value to a global `void *` and crash if it changes from one non-zero value to another. Well, formally speaking it would require `<atomic>`, which is also a C++11 thing, but one can make "best effort" with volatiles...

Reply:
Yes, it is atrocious that Visual C++ does not respect even getting `__cplusplus` correct.